add_library cannot create imported target "ffmpeg::libavcodec"

<!--- Please fill out this template to the best of your ability. You can always edit this issue once you have created it. -->
<!--- Read the following link before you create a new problem report: https://kodi.wiki/view/HOW-TO:Submit_a_bug_report  -->
## Bug report
### Describe the bug
Here is a clear and concise description of what the problem is: Error building last stable release
<!--- Provide a more detailed introduction to the issue itself, and why you consider it to be a bug -->
<!--- A bug report that is not clear will be closed -->
<!--- Put your text below this line -->
`cmake` can not complete the configure process due to some `ffmpeg` errors: a missing FFmpeg 6.0.0 software and a duplicated "ffmpeg::libavcodec" target.


## Expected Behavior
Here is a clear and concise description of what was expected to happen:
<!--- Tell us what should happen -->
<!--- Put your text below this line -->
The configure process should ends with a message: `-- Build files have been written to: /usr/local/share/kodi-build`


## Actual Behavior
<!--- Tell us what happens instead -->
<!--- Put your text below this line -->



## Possible Fix
<!--- Not obligatory, but suggest a fix or reason for the bug -->
<!--- Put your text below this line -->



### To Reproduce
Steps to reproduce the behavior:
<!--- Provide a link to a live example, or an unambiguous set of steps to -->
<!--- reproduce this bug. Include code to reproduce, if relevant -->
<!--- Put your text below this line -->
1. cd /usr/local/share/kodi
2. sudo make -C tools/depends/target/crossguid PREFIX=/usr/local
3. sudo make -C tools/depends/target/flatbuffers PREFIX=/usr/local
4. sudo make -C tools/depends/target/fmt PREFIX=/usr/local
5. sudo make -C tools/depends/target/spdlog PREFIX=/usr/local
6. sudo make -C tools/depends/target/wayland-protocols PREFIX=/usr/local
7. sudo make -C tools/depends/target/waylandpp PREFIX=/usr/local
8. mkdir /usr/local/share/kodi-build
9. cmake ../kodi  -DCMAKE_INSTALL_PREFIX=/usr/local  -DCORE_PLATFORM_NAME=x11 -DAPP_RENDER_SYSTEM=gl -DENABLE_INTERNAL_FLATBUFFERS=ON 


### Debuglog
<!--- Put your text below this line -->
<!--- A debuglog is always mandatory when creating an issue. Provide one! -->
The debuglog can be found here:

[kodi_cmake.log](https://github.com/xbmc/xbmc/files/13641764/kodi_cmake.log)


### Screenshots 
Here are some links or screenshots to help explain the problem:
<!--- Put your text below this line -->



## Additional context or screenshots (if appropriate)
Here is some additional context or explanation that might help:
<!--- How has this bug affected you? What were you trying to accomplish? -->
<!--- Put your text below this line -->

The OS already has the `FFmpeg 6.0.0` fetched from deb-multimedia repository.
```
root@slimshadygray:~# dpkg -l ffmpeg*
Desired=Unknown/Install/Remove/Purge/Hold
| Status=Not/Inst/Conf-files/Unpacked/halF-conf/Half-inst/trig-aWait/Trig-pend
|/ Err?=(none)/Reinst-required (Status,Err: uppercase=bad)
||/ Name              Version                     Architecture Description
+++-=================-===========================-============-==================================================================
ii  ffmpeg            10:6.0.1-dmo0+deb12u1       amd64        audio/video encoder, streaming server & audio/video file converter
ii  ffmpeg-doc        7:5.1.4-0+deb12u1           all          Documentation of the FFmpeg multimedia framework
ii  ffmpegthumbnailer 2.2.2+git20220218+dfsg-1+b1 amd64        fast and lightweight video thumbnailer
ii  ffmpegthumbs      4:22.12.3-1                 amd64        video thumbnail generator using ffmpeg
root@slimshadygray:~#
```
So, it doesn't make sense the complain `FFmpeg 6.0.0 not found, falling back to internal build`

### Your Environment
Used Operating system:
<!--- Include as many relevant details about the environment you experienced the bug in -->
<!--- Put your text below this line. Checkboxes can easily be ticked once issue is created -->
 - [ ] Android
 - [ ] iOS
 - [ ] tvOS
 - [ ] Linux
 - [ ] macOS
 - [ ] Windows
 - [ ] Windows UWP

 - Operating system version/name: Debian GNU/Linux 12.4 «Bookworm»
 - Kodi version: 21.0 BETA2 20.90.821



<!--- End of this issue -->
*note: Once the issue is made we require you to update it with new information or Kodi versions should that be required.
Team Kodi will consider your problem report however, we will not make any promises the problem will be solved.*

I can reproduce on Ubuntu 22.04. If I clone Kodi and do cmake in /build, it works. But if I've used depends to build a package (I need to build libdisplay-info from source), then I get the error above. Specifically, it happens in my build script on [this line](https://github.com/eigendude/OASIS/blob/main/oasis_tooling/scripts/build_kodi.sh#L134) if I comment out `-DENABLE_INTERNAL_FFMPEG=ON`. The error is:

```

-- FFmpeg 6.0.0 not found, falling back to internal build
-- Could NOT find Dav1d (missing: DAV1D_LIBRARY DAV1D_INCLUDE_DIR) (Required is at least version "1.1.0")
-- dav1d not found, internal ffmpeg build will be missing AV1 support!
CMake Error at cmake/modules/FindFFMPEG.cmake:128 (add_library):
  add_library cannot create imported target "ffmpeg::libavcodec" because
  another target with the same name already exists.
Call Stack (most recent call first):
  cmake/modules/FindFFMPEG.cmake:381 (buildFFMPEG)
  cmake/scripts/common/Macros.cmake:388 (find_package)
  cmake/scripts/common/Macros.cmake:402 (find_package_with_ver)
  CMakeLists.txt:257 (core_require_dep)
```

--- a/cmake/modules/FindFFMPEG.cmake
+++ b/cmake/modules/FindFFMPEG.cmake
@@ -195,156 +195,10 @@ else()
                   libswresample${_swresample_ver}
                   libpostproc${_postproc_ver})
 
-  if(PKG_CONFIG_FOUND)
-    pkg_check_modules(PC_FFMPEG ${FFMPEG_PKGS} QUIET)
-  endif()
+  if(NOT WIN32)
+    find_package(PkgConfig REQUIRED)
 
-  find_path(FFMPEG_INCLUDE_DIRS libavcodec/avcodec.h libavfilter/avfilter.h libavformat/avformat.h
-                                libavutil/avutil.h libswscale/swscale.h libpostproc/postprocess.h
-            PATH_SUFFIXES ffmpeg
-            PATHS ${PC_FFMPEG_INCLUDE_DIRS}
-            NO_DEFAULT_PATH)
-  find_path(FFMPEG_INCLUDE_DIRS libavcodec/avcodec.h libavfilter/avfilter.h libavformat/avformat.h
-                                libavutil/avutil.h libswscale/swscale.h libpostproc/postprocess.h)
-
-  find_library(FFMPEG_LIBAVCODEC
-               NAMES avcodec libavcodec
-               PATH_SUFFIXES ffmpeg/libavcodec
-               PATHS ${PC_FFMPEG_libavcodec_LIBDIR}
-               NO_DEFAULT_PATH)
-  find_library(FFMPEG_LIBAVCODEC NAMES avcodec libavcodec PATH_SUFFIXES ffmpeg/libavcodec)
-
-  if(FFMPEG_LIBAVCODEC)
-    if(PKG_CONFIG_FOUND)
-      pkg_check_modules(PC_FFMPEG_LIB libavcodec${_avcodec_ver} QUIET)
-      string(REGEX REPLACE "framework;" "framework " PC_FFMPEG_LIB_LDFLAGS "${PC_FFMPEG_LIB_LDFLAGS}")
-    endif()
-    add_library(ffmpeg::libavcodec UNKNOWN IMPORTED)
-    set_target_properties(ffmpeg::libavcodec PROPERTIES
-                                         FOLDER "FFMPEG - External Projects"
-                                         IMPORTED_LOCATION "${FFMPEG_LIBAVCODEC}"
-                                         INTERFACE_LINK_LIBRARIES "${PC_FFMPEG_LIB_LDFLAGS}"
-                                         INTERFACE_INCLUDE_DIRECTORIES "${FFMPEG_INCLUDE_DIRS}")
-  endif()
-
-  find_library(FFMPEG_LIBAVFILTER
-               NAMES avfilter libavfilter
-               PATH_SUFFIXES ffmpeg/libavfilter
-               PATHS ${PC_FFMPEG_libavfilter_LIBDIR}
-               NO_DEFAULT_PATH)
-  find_library(FFMPEG_LIBAVFILTER NAMES avfilter libavfilter PATH_SUFFIXES ffmpeg/libavfilter)
-
-  if(FFMPEG_LIBAVFILTER)
-    if(PKG_CONFIG_FOUND)
-      pkg_check_modules(PC_FFMPEG_LIB libavfilter${_avfilter_ver} QUIET)
-      string(REGEX REPLACE "framework;" "framework " PC_FFMPEG_LIB_LDFLAGS "${PC_FFMPEG_LIB_LDFLAGS}")
-    endif()
-    add_library(ffmpeg::libavfilter UNKNOWN IMPORTED)
-    set_target_properties(ffmpeg::libavfilter PROPERTIES
-                                         FOLDER "FFMPEG - External Projects"
-                                         IMPORTED_LOCATION "${FFMPEG_LIBAVFILTER}"
-                                         INTERFACE_LINK_LIBRARIES "${PC_FFMPEG_LIB_LDFLAGS}"
-                                         INTERFACE_INCLUDE_DIRECTORIES "${FFMPEG_INCLUDE_DIRS}")
-  endif()
-
-  find_library(FFMPEG_LIBAVFORMAT
-               NAMES avformat libavformat
-               PATH_SUFFIXES ffmpeg/libavformat
-               PATHS ${PC_FFMPEG_libavformat_LIBDIR}
-               NO_DEFAULT_PATH)
-  find_library(FFMPEG_LIBAVFORMAT NAMES avformat libavformat PATH_SUFFIXES ffmpeg/libavformat)
-
-  if(FFMPEG_LIBAVFORMAT)
-    if(PKG_CONFIG_FOUND)
-      pkg_check_modules(PC_FFMPEG_LIB libavformat${_avformat_ver} QUIET)
-      string(REGEX REPLACE "framework;" "framework " PC_FFMPEG_LIB_LDFLAGS "${PC_FFMPEG_LIB_LDFLAGS}")
-    endif()
-    add_library(ffmpeg::libavformat UNKNOWN IMPORTED)
-    set_target_properties(ffmpeg::libavformat PROPERTIES
-                                         FOLDER "FFMPEG - External Projects"
-                                         IMPORTED_LOCATION "${FFMPEG_LIBAVFORMAT}"
-                                         INTERFACE_LINK_LIBRARIES "${PC_FFMPEG_LIB_LDFLAGS}"
-                                         INTERFACE_INCLUDE_DIRECTORIES "${FFMPEG_INCLUDE_DIRS}")
-  endif()
-
-  find_library(FFMPEG_LIBAVUTIL
-               NAMES avutil libavutil
-               PATH_SUFFIXES ffmpeg/libavutil
-               PATHS ${PC_FFMPEG_libavutil_LIBDIR}
-               NO_DEFAULT_PATH)
-  find_library(FFMPEG_LIBAVUTIL NAMES avutil libavutil PATH_SUFFIXES ffmpeg/libavutil)
-
-  if(FFMPEG_LIBAVUTIL)
-    if(PKG_CONFIG_FOUND)
-      pkg_check_modules(PC_FFMPEG_LIB libavutil${_avutil_ver} QUIET)
-      string(REGEX REPLACE "framework;" "framework " PC_FFMPEG_LIB_LDFLAGS "${PC_FFMPEG_LIB_LDFLAGS}")
-    endif()
-    add_library(ffmpeg::libavutil UNKNOWN IMPORTED)
-    set_target_properties(ffmpeg::libavutil PROPERTIES
-                                         FOLDER "FFMPEG - External Projects"
-                                         IMPORTED_LOCATION "${FFMPEG_LIBAVUTIL}"
-                                         INTERFACE_LINK_LIBRARIES "${PC_FFMPEG_LIB_LDFLAGS}"
-                                         INTERFACE_INCLUDE_DIRECTORIES "${FFMPEG_INCLUDE_DIRS}")
-  endif()
-
-  find_library(FFMPEG_LIBSWSCALE
-               NAMES swscale libswscale
-               PATH_SUFFIXES ffmpeg/libswscale
-               PATHS ${PC_FFMPEG_libswscale_LIBDIR}
-               NO_DEFAULT_PATH)
-  find_library(FFMPEG_LIBSWSCALE NAMES swscale libswscale PATH_SUFFIXES ffmpeg/libswscale)
-
-  if(FFMPEG_LIBSWSCALE)
-    if(PKG_CONFIG_FOUND)
-      pkg_check_modules(PC_FFMPEG_LIB libswscale${_swscale_ver} QUIET)
-      string(REGEX REPLACE "framework;" "framework " PC_FFMPEG_LIB_LDFLAGS "${PC_FFMPEG_LIB_LDFLAGS}")
-    endif()
-    add_library(ffmpeg::libswscale UNKNOWN IMPORTED)
-    set_target_properties(ffmpeg::libswscale PROPERTIES
-                                         FOLDER "FFMPEG - External Projects"
-                                         IMPORTED_LOCATION "${FFMPEG_LIBSWSCALE}"
-                                         INTERFACE_LINK_LIBRARIES "${PC_FFMPEG_LIB_LDFLAGS}"
-                                         INTERFACE_INCLUDE_DIRECTORIES "${FFMPEG_INCLUDE_DIRS}")
-  endif()
-
-  find_library(FFMPEG_LIBSWRESAMPLE
-               NAMES swresample libswresample
-               PATH_SUFFIXES ffmpeg/libswresample
-               PATHS ${PC_FFMPEG_libswresample_LIBDIR}
-               NO_DEFAULT_PATH)
-  find_library(FFMPEG_LIBSWRESAMPLE NAMES NAMES swresample libswresample PATH_SUFFIXES ffmpeg/libswresample)
-
-  if(FFMPEG_LIBSWRESAMPLE)
-    if(PKG_CONFIG_FOUND)
-      pkg_check_modules(PC_FFMPEG_LIB libswresample${_swresample_ver} QUIET)
-      string(REGEX REPLACE "framework;" "framework " PC_FFMPEG_LIB_LDFLAGS "${PC_FFMPEG_LIB_LDFLAGS}")
-    endif()
-    add_library(ffmpeg::libswresample UNKNOWN IMPORTED)
-    set_target_properties(ffmpeg::libswresample PROPERTIES
-                                         FOLDER "FFMPEG - External Projects"
-                                         IMPORTED_LOCATION "${FFMPEG_LIBSWRESAMPLE}"
-                                         INTERFACE_LINK_LIBRARIES "${PC_FFMPEG_LIB_LDFLAGS}"
-                                         INTERFACE_INCLUDE_DIRECTORIES "${FFMPEG_INCLUDE_DIRS}")
-  endif()
-
-  find_library(FFMPEG_LIBPOSTPROC
-               NAMES postproc libpostproc
-               PATH_SUFFIXES ffmpeg/libpostproc
-               PATHS ${PC_FFMPEG_libpostproc_LIBDIR}
-               NO_DEFAULT_PATH)
-  find_library(FFMPEG_LIBPOSTPROC NAMES postproc libpostproc PATH_SUFFIXES ffmpeg/libpostproc)
-
-  if(FFMPEG_LIBPOSTPROC)
-    if(PKG_CONFIG_FOUND)
-      pkg_check_modules(PC_FFMPEG_LIB libpostproc${_postproc_ver} QUIET)
-      string(REGEX REPLACE "framework;" "framework " PC_FFMPEG_LIB_LDFLAGS "${PC_FFMPEG_LIB_LDFLAGS}")
-    endif()
-    add_library(ffmpeg::libpostproc UNKNOWN IMPORTED)
-    set_target_properties(ffmpeg::libpostproc PROPERTIES
-                                         FOLDER "FFMPEG - External Projects"
-                                         IMPORTED_LOCATION "${FFMPEG_LIBPOSTPROC}"
-                                         INTERFACE_LINK_LIBRARIES "${PC_FFMPEG_LIB_LDFLAGS}"
-                                         INTERFACE_INCLUDE_DIRECTORIES "${FFMPEG_INCLUDE_DIRS}")
+    pkg_check_modules(PC_FFMPEG ${FFMPEG_PKGS})
   endif()
 
   if((PC_FFMPEG_FOUND
@@ -358,6 +212,29 @@ else()
      OR WIN32)
     set(FFMPEG_VERSION ${REQUIRED_FFMPEG_VERSION})
 
+    # macro for find_library usage
+    # arg1: lowercase libname (eg libavcodec, libpostproc, etc)
+    macro(ffmpeg_find_lib libname)
+      string(TOUPPER ${libname} libname_UPPER)
+      string(REPLACE "lib" "" name ${libname})
+
+      find_library(FFMPEG_${libname_UPPER}
+                   NAMES ${name} ${libname}
+                   PATH_SUFFIXES ffmpeg/${libname}
+                   HINTS ${DEPENDS_PATH}/lib ${PC_FFMPEG_${libname}_LIBDIR}
+                   ${${CORE_PLATFORM_LC}_SEARCH_CONFIG})
+    endmacro()
+
+    find_path(FFMPEG_INCLUDE_DIRS libavcodec/avcodec.h libavfilter/avfilter.h libavformat/avformat.h
+                                  libavutil/avutil.h libswscale/swscale.h libpostproc/postprocess.h
+              PATH_SUFFIXES ffmpeg
+              HINTS ${DEPENDS_PATH}/include
+              ${${CORE_PLATFORM_LC}_SEARCH_CONFIG})
+
+    foreach(_ffmpeg_pkg IN ITEMS ${FFMPEG_PKGS})
+      string(REGEX REPLACE ">=.*" "" _libname ${_ffmpeg_pkg})
+      ffmpeg_find_lib(${_libname})
+    endforeach()
 
     include(FindPackageHandleStandardArgs)
     find_package_handle_standard_args(FFMPEG
@@ -372,6 +249,44 @@ else()
                                                     FFMPEG_LIBPOSTPROC
                                                     FFMPEG_VERSION
                                       FAIL_MESSAGE "FFmpeg ${REQUIRED_FFMPEG_VERSION} not found, please consider using -DENABLE_INTERNAL_FFMPEG=ON")
+
+    # Macro to populate target
+    # arg1: lowercase libname (eg libavcodec, libpostproc, etc)
+    macro(ffmpeg_create_target libname)
+      string(TOUPPER ${libname} libname_UPPER)
+      string(REPLACE "lib" "" name ${libname})
+
+      if(PKG_CONFIG_FOUND AND NOT WIN32)
+        # We have to run the check against the single lib a second time, as when
+        # pkg_check_modules is run with a list, the only *_LDFLAGS set is a concatenated 
+        # list of all checked modules. Ideally we want each target to only have the LDFLAGS
+        # required for that specific module
+        pkg_check_modules(PC_FFMPEG_${libname} ${libname}${_${name}_ver} QUIET)
+
+        # pkg-config LDFLAGS always seem to have -l<name> listed. We dont need that, as
+        # the target gets a direct path to the physical lib
+        list(REMOVE_ITEM PC_FFMPEG_${libname}_LDFLAGS "-l${name}")
+
+        # Darwin platforms return a list that cmake splits "framework libname" into separate
+        # items, therefore the link arguments become -framework -libname causing link failures
+        # we just force concatenation of these instances, so cmake passes it as "-framework libname"
+        if(CMAKE_SYSTEM_NAME STREQUAL "Darwin")
+          string(REGEX REPLACE "framework;" "framework " PC_FFMPEG_${libname}_LDFLAGS "${PC_FFMPEG_${libname}_LDFLAGS}")
+        endif()
+      endif()
+
+      add_library(ffmpeg::${libname} UNKNOWN IMPORTED)
+      set_target_properties(ffmpeg::${libname} PROPERTIES
+                                           FOLDER "FFMPEG - External Projects"
+                                           IMPORTED_LOCATION "${FFMPEG_${libname_UPPER}}"
+                                           INTERFACE_LINK_LIBRARIES "${PC_FFMPEG_${libname}_LDFLAGS}"
+                                           INTERFACE_INCLUDE_DIRECTORIES "${FFMPEG_INCLUDE_DIRS}")
+    endmacro()
+
+    foreach(_ffmpeg_pkg IN ITEMS ${FFMPEG_PKGS})
+      string(REGEX REPLACE ">=.*" "" _libname ${_ffmpeg_pkg})
+      ffmpeg_create_target(${_libname})
+    endforeach()
 
   else()
     if(FFMPEG_PATH)
@@ -407,5 +322,3 @@ if(FFMPEG_FOUND)
 
   set_property(GLOBAL APPEND PROPERTY INTERNAL_DEPS_PROP ffmpeg::ffmpeg)
 endif()
-
-mark_as_advanced(FFMPEG_FOUND)

